rfcs/20190814-kernel-and-op-registration.md

```diff
@@ -0,0 +1,658 @@
+
+## Kernels
+
+This section introduces changes to the C API that are required to support
+kernels. An alpha version of this API is already checked in at
+`tensorflow/c/kernels.h`.
+
+### Registration
+
+Kernel registration with the C++ API is accomplished with the
+`REGISTER_KERNEL_BUILDER` macro. This macro expands to code that relies on
+static initialization to register the provided kernel with the global kernel
+registry. See below for an example of registering a kernel with the C++ API:
+
+```c++
+
+#include "tensorflow/core/framework/op_kernel.h"
+
+class BitcastOp : public OpKernel {
+  explicit BitcastOp(OpKernelConstruction* context) : OpKernel(context) { … }
+  void Compute(OpKernelContext* context) override { … }
+};
+
+REGISTER_KERNEL_BUILDER(Name("Bitcast").Device(DEVICE_CPU), BitcastOp)
+```
+
+The equivalent C API provides a series of functions that operate on
+`TF_KernelBuilder`, an opaque struct obtained with the `TF_NewKernelBuilder` call.
+The kernel builder is registered with TensorFlow using the
+`TF_RegisterKernelBuilder` function. See below for an example of registering
+the bitcast kernel using the C API:
+
+```c++
+#include "tensorflow/c/kernels.h"
+
+typedef struct bitcast_kernel { … } bitcast_kernel;
+
+// Bitcast_Create, Bitcast_Compute and Bitcast_Delete actually implement the
+// kernel. See the section below for discussion on kernel implementation.
+static void* Bitcast_Create(TF_OpKernelConstruction* context) {
+  bitcast_kernel* k = (bitcast_kernel*) calloc(1, sizeof(bitcast_kernel));
+  /* initialize the fields of k as needed */
+  return (void*) k;
+}
+
+static void* Bitcast_Compute(void* k, TF_OpKernelContext* context) {
+  bitcast_kernel* kernel = (bitcast_kernel*) k;  // this is the pointer returned by
+                                                 // Bitcast_Create
+  /* compute the result */
+  TF_SetOutput(context, ...);
+}
+
+static void Bitcast_Delete(void *k) { free(k); }
+
+void InitPlugin() {
+  TF_KernelBuilder* builder = TF_NewKernelBuilder(/*op_name*/"Bitcast", DEVICE_CPU,
+      &Bitcast_Create, &Bitcast_Compute, &Bitcast_Delete);
+  TF_Status* status = TF_NewStatus();
+  TF_RegisterKernelBuilder(/*kernel_name*/"Bitcast", builder, status);
+  if (TF_GetCode(status) != TF_OK) { /* handle errors */ }
+  TF_DeleteStatus(status);
+}
+```
+
+The registration function prototypes are provided below. Kernel authors must
+provide a compute function. Creation and deletion functions are optional, but
+if a creation function is provided that causes memory allocation, a deletion
+function that frees the memory should also be provided, otherwise a leak will
+occur.
+
+```c++
+TF_KernelBuilder* TF_NewKernelBuilder(
+  const char* op_name, const char* device_name,
+  void* (*create_func)(TF_OpKernelConstruction*),
+  void (*compute_func)(void*, TF_OpKernelContext*),
+  void (*delete_func)(void*));
+
+void TF_RegisterKernelBuilder(const char* name, TF_KernelBuilder* builder,
```

Comment:
Why two functions instead of one function?

Reply:
The kernel builder may be customized with other calls, such as `TF_KernelBuilder_TypeConstraint` and `TF_KernelBuilder_HostMemory`, I'll add these to the doc.